### PR TITLE
[resubmit][rpc] Add local shutdown to process group agent

### PR DESCRIPTION
**Pull request by rohan-varma**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#30330 [resubmit][rpc] Add local shutdown to process group agent**
[test all]
Resubmit of https://github.com/pytorch/pytorch/pull/30020, which was reverted.
This is now possible due to previous changes made in `gloo` and `ProcessGroupGloo`. We `abort` the listener thread that is waiting for a message, and join all other threads. The destructor calls this same `localShutdown` method, but we ensure this is not called multiple times.

Differential Revision: [D18661775](https://our.internmc.facebook.com/intern/diff/D18661775/)

### CODE REVIEW

File: test/rpc_test.py
Comment (mrshenli) [inline test/rpc_test.py]: IIUC, `self. rpc_agent_options` no longer exists, this needs to be replaced by  `rpc_backend_options`?

Comment (rohan-varma) [reply test/rpc_test.py]: Sorry, looks like the update was coming just as you reviewed

**Comment by zhaojuanmao** [comment]
@rohan-varma and @mrshenli I'm wondering why we need to split into wait_all_workers() and shutdown() in this way if shutdown() always needs to follow by wait_all_workers()?  looks like they are not independent.

**Comment by mrshenli** [comment]
@zhaojuanmao 

One reason is that `wait_all_workers` could become a non-terminating API in the future, sth similar to `dist.barrier()`. So that applications can do some rpc/remote and then call `wait_all_workers` for a global sync, and then resume, and then maybe another global sync, and then `shutdown`. (as mentioned by @satgera) We don't explicitly say we support this yet, but we can discuss if that's what we want.

File: torch/csrc/distributed/rpc/process_group_agent.cpp
Comment (rohan-varma) [inline torch/csrc/distributed/rpc/process_group_agent.cpp]: I had to add this here in shutdown, otherwise the test `test_rpc_shutdown()` would flake with gloo connection reset errors. I don't think it delays the shutdown by too much, as it just waits for already enqueued work to complete.

**Comment by zhaojuanmao** [comment]
lgtm!

File: torch/distributed/rpc/api.py
Comment (xush6528) [inline torch/distributed/rpc/api.py]: "RPC handlers" -> "Python RPC handlers"


RPC handlers is vague.

Comment (rohan-varma) [reply torch/distributed/rpc/api.py]: This docstring is also in the wrong place now, will update

Comment (mrshenli) [reply torch/distributed/rpc/api.py]: > I actually don't think we need to mention that we destroy RPC handlers and RRef context, those are internal details that should be abstracted away from the user.

Agree, users don't need to know there is a RRef context and handler

File: torch/distributed/rpc/api.py
Comment (mrshenli) [inline torch/distributed/rpc/api.py]: (This can come in a followup PR)

We might not need this `global _agent` any more as we are not modifying it.

File: torch/csrc/distributed/rpc/process_group_agent.cpp
Comment (mrshenli) [reply torch/csrc/distributed/rpc/process_group_agent.cpp]: hmm, does this need to be placed before abort? Otherwise, if there are indeed unfinished task, the receiving end listener thread might have already been aborted?

**Review by mrshenli — APPROVED** [review]
Approve to unblock. Please wait for all tests to pass.

File: torch/csrc/distributed/rpc/process_group_agent.cpp
Comment (xush6528) [inline torch/csrc/distributed/rpc/process_group_agent.cpp]: What if the `futureTimeoutThread_` is not blocked on the `futureTimeoutCV_`? For example, it could be at line `std::chrono::milliseconds sleepTime;`.

Will `futureTimeoutThread_` live forever?

File: torch/csrc/distributed/rpc/process_group_agent.cpp
Comment (xush6528) [inline torch/csrc/distributed/rpc/process_group_agent.cpp]: I changed the `futureTimeoutCV_.wait_for(...)` part in https://github.com/pytorch/pytorch/pull/30355.

We need to change this line to
```
std::lock_guar<std::mutex> futureLock{futureMutex_};
rpcRunning_.store(true);
```
in #30355 as well.

File: torch/csrc/distributed/rpc/process_group_agent.cpp
Comment (xush6528) [inline torch/csrc/distributed/rpc/process_group_agent.cpp]: Here as well.

```cpp
std::lock_guar<std::mutex> futureLock{futureMutex_};
if (!rpcRunning_.exchange(false)) {
  return;
}
```

**Review by mrshenli — CHANGES_REQUESTED** [review]
Remove stamp until we address @xush6528's comment

File: torch/csrc/distributed/rpc/process_group_agent.cpp
Comment (rohan-varma) [reply torch/csrc/distributed/rpc/process_group_agent.cpp]: Hm, my reasoning for doing it after was that if it was before, we could possibly get more work while aborting the listenerThread, and then enqueue that recv, and not wait for it to be completed. For example, say that recvWork->wait has already been unblocked before we call its abort (via a message from another worker), we may have called enqueueRecv which will add a task to the thread pool.

File: torch/csrc/distributed/rpc/process_group_agent.cpp
Comment (rohan-varma) [reply torch/csrc/distributed/rpc/process_group_agent.cpp]: Thanks for pointing this out - so it sounds like this should be fixed if we hold the lock on `rpcRunning`? Should we hold that lock wherever we write/read it?

File: torch/csrc/distributed/rpc/process_group_agent.cpp
Comment (rohan-varma) [reply torch/csrc/distributed/rpc/process_group_agent.cpp]: Added.

File: torch/csrc/distributed/rpc/process_group_agent.cpp
Comment (rohan-varma) [inline torch/csrc/distributed/rpc/process_group_agent.cpp]: We now hold the lock from when we check `rpcRunning` to until we sleep, and the lock is also held in `shutdown()` before we set `rpcRunning`.

File: torch/csrc/distributed/rpc/process_group_agent.cpp
Comment (mrshenli) [inline torch/csrc/distributed/rpc/process_group_agent.cpp]: Just make sure I understand this correctly. If this is before the setting `rpcRunning_` in `shutdown`, then `futureTimeoutCV_` will be notified properly. If this is after, then this would return as `rpcRunning_.load() == false`, right?

**Review by mrshenli — APPROVED** [review]
Thanks for fixing promptly!

File: torch/csrc/distributed/rpc/process_group_agent.cpp
Comment (rohan-varma) [inline torch/csrc/distributed/rpc/process_group_agent.cpp]: This check is not strictly necessary, but can help performance. If the `shutdown` thread wakes up the CV here, then `rpcRunning` will always be false, without this check, we would continue to `processTimedOutFutures`, which would be a no-op.

File: test/rpc_test.py
Comment (pietern) [inline test/rpc_test.py]: This can do `time.sleep(0)` to yield control and effectively busy wait.

File: torch/csrc/distributed/rpc/process_group_agent.h
Comment (pietern) [inline torch/csrc/distributed/rpc/process_group_agent.h]: Since you're locking access anyway and not using this as an atomic, I recommend changing this to a regular bool to avoid the impression you can use it as atomic.

File: torch/csrc/distributed/rpc/process_group_agent.cpp
Comment (xush6528) [inline torch/csrc/distributed/rpc/process_group_agent.cpp]: I am not convinced yet. How do you make sure the notification is always delivered with no cond var predicate added?

e.g. What if `pollTimedOutRPCs` is at here, and `shutdown` sends a notification? Would that notification be lost?

**Review by xush6528 — APPROVED** [review]

File: torch/csrc/distributed/rpc/process_group_agent.cpp
Comment (rohan-varma) [reply torch/csrc/distributed/rpc/process_group_agent.cpp]: For `shutdown` to have sent a notification, this implies that it would have had to hold the `futureMutex` lock and set `rpcRunning` to be false before. But since that now requires the lock, we can't have `shutdown` be at `std::chrono::milliseconds sleepTime;` at the same time, since this section is also locked.

File: torch/csrc/distributed/rpc/process_group_agent.h
Comment (rohan-varma) [reply torch/csrc/distributed/rpc/process_group_agent.h]: Synced with @xush6528 about this - currently, it is not always locked (see the checks in `listenLoop` and `send`). We only need to strictly lock it when it is used in conjunction with the `futureTimeout` condvar, so it might be worth leaving it as atomic?

File: torch/csrc/distributed/rpc/process_group_agent.cpp
Comment (jjlilley) [inline torch/csrc/distributed/rpc/process_group_agent.cpp]: why do you need to hold the mutex across pg_->recvAnySource()?

Comment (rohan-varma) [reply torch/csrc/distributed/rpc/process_group_agent.cpp]: We want to make sure that `recvWork` is in a valid initialized state when we call `abort` on it from the shutdown thread, so we'd want the lock to ensure there is no race when it is initialized here and when it is checked in the `shutdown` function.

Comment (jjlilley) [reply torch/csrc/distributed/rpc/process_group_agent.cpp]: Ah, I see - I initially didn't see that this was a separate mutex. Thanks.

fwiw, it's a little sketchy to access recvWork_ on 479 without holding the mutex, you should ideally keep a local ref to the shared_ptr, e.g.

auto work = pg_->recvAnysource(preamble, pg_->getRank());
setPendingRecvWork(work);  // or put inline, just grabs lock and sets recvWork_
if (!rpcRunning_.load() ||
   !work->wait() /* aborted */) {
   return;
}

If nothing else, it's easier to reason about shutdown safety, when we guarantee that recvWork_ is only accessed under the lock.

File: torch/csrc/distributed/rpc/process_group_agent.cpp
Comment (rohan-varma) [reply torch/csrc/distributed/rpc/process_group_agent.cpp]: Yes - for the first case you mentioned, this thread would drop the lock when it waits and would be appropriately notified by shutdown. For the 2nd case, we would break as `rpcRunning = false`.

File: torch/csrc/distributed/rpc/process_group_agent.cpp
Comment (rohan-varma) [reply torch/csrc/distributed/rpc/process_group_agent.cpp]: Thanks for calling this out, this is implemented in the latest version.

Comment (jjlilley) [reply torch/csrc/distributed/rpc/process_group_agent.cpp]: Great!
fwiw, LGTM

File: torch/distributed/rpc/api.py
Comment (mrshenli) [inline torch/distributed/rpc/api.py]: When you rebase onto #30217, could you please please move this line to the front of `_agent.shutdown()`. This shouldn't make any difference in our tests for now. However, `ctx.destroyed_` is only set to `true` in `_destroy_rref_context()`, meaning that logically, RRef destructions could still trigger messages been sent using the `_agent` before `_destroy_rref_context()`.

Comment (rohan-varma) [reply torch/distributed/rpc/api.py]: Should it be before `_wait_all_workers` as well then, since we want to wait for these messages to be processed?

Comment (mrshenli) [reply torch/distributed/rpc/api.py]: Good point! Yes, it should!

**Comment by rohan-varma** [comment]
cc @pritamdamania87, this is the doc update: 
doc_update

File: torch/distributed/rpc/api.py
Comment (rohan-varma) [inline torch/distributed/rpc/api.py]: cc @mrshenli this is the new shutdown sequence after the ignore rref leaks PR has landed

Comment (mrshenli) [reply torch/distributed/rpc/api.py]: this looks good to me, thanks!

**Review by mrshenli — APPROVED** [review]

File: docs/source/notes/distributed_autograd.rst
Comment (pietern) [inline docs/source/notes/distributed_autograd.rst]: `mp.spawn` for the win!

File: test/rpc_test.py
Comment (pietern) [inline test/rpc_test.py]: Could add a comment here to say that this yields control to other threads so it won't be removed.

File: torch/csrc/distributed/rpc/process_group_agent.h
Comment (pietern) [reply torch/csrc/distributed/rpc/process_group_agent.h]: Sounds good, but please document what the expected locking strategy is.

Using an atomic implies (at least to me) that you don't need locks. To reduce the probability of this causing breakage later on, I think a regular bool is better. Accessing that without locks is a red flag in code review.

File: test/rpc_test.py
Comment (rohan-varma) [reply test/rpc_test.py]: Will address in a follow up PR.

File: docs/source/notes/distributed_autograd.rst
Comment (rohan-varma) [reply docs/source/notes/distributed_autograd.rst]: Addressing in separate PR (https://github.com/pytorch/pytorch/pull/30381)

**Comment by rohan-varma** [comment]
The following are the CI errors, all unrelated:

1) Error response from daemon: manifest for 308535385114.dkr.ecr.us-east-1.amazonaws.com/pytorch/pytorch-linux-xenial-py3-clang5-android-ndk-r19c:405-3796e2c20d4587ffdcae63b42490a63316be44bd-android-arm-v7a not found

2) Nov 27 08:55:40 FAIL [0.032s]: test_neg_cuda (__main__.TestTorchDeviceTypeCUDA)
Nov 27 08:55:40 ----------------------------------------------------------------------
Nov 27 08:55:40 RuntimeError: CUDA error: misaligned address

3) Nov 27 00:01:04 Traceback (most recent call last):
Nov 27 00:01:04   File "/Users/distiller/workspace/miniconda3/lib/python3.7/site-packages/soundfile.py", line 142, in <module>
Nov 27 00:01:04     raise OSError('sndfile library not found')
Nov 27 00:01:04 OSError: sndfile library not found

4) 08:34:39 FAIL: test_logical_or_cuda (__main__.TestTorchDeviceTypeCUDA)
08:34:39 ----------------------------------------------------------------------
08:34:39 Traceback (most recent call last):
08:34:39   File "/var/lib/jenkins/workspace/test/common_utils.py", line 631, in wrapper
08:34:39     method(*args, **kwargs)
08:34:39   File "/var/lib/jenkins/workspace/test/common_device_type.py", line 179, in instantiated_test
08:34:39     return test(self, device_arg)
08:34:39   File "test_torch.py", line 6578, in test_logical_or
08:34:39     self._test_logical(device, 'logical_or', [10, 0, 1, 0], [1, 0, 0, 10], [1, 0, 1, 1])
08:34:39   File "test_torch.py", line 6555, in _test_logical
08:34:39     self.assertEqual(expected_res.bool(), getattr(a, op)(b))
08:34:39   File "/var/lib/jenkins/workspace/test/common_utils.py", line 808, in assertEqual
08:34:39     assertTensorsEqual(x, y)
08:34:39   File "/var/lib/jenkins/workspace/test/common_utils.py", line 778, in assertTensorsEqual
08:34:39     self.assertLessEqual(max_err, prec, message)
08:34:39 AssertionError: tensor(1, device='cuda:0', dtype=torch.int32) not less than or equal to 1e-05 :

File: torch/csrc/distributed/rpc/process_group_agent.h
Comment (rohan-varma) [reply torch/csrc/distributed/rpc/process_group_agent.h]: Will add this in a follow up PR.

**Comment by rohan-varma** [comment]
CI failure that is not listed as "broken upstream" has occured on other merged PRs (https://app.circleci.com/jobs/github/pytorch/pytorch/3733973,  https://app.circleci.com/jobs/github/pytorch/pytorch/3734258), it should be fixed by this revert: https://github.com/pytorch/pytorch/commit/a2ed50c9209cb61eb72465b55936c4bdc468cc35

File: test/rpc_test.py
Comment (rohan-varma) [reply test/rpc_test.py]: I ended up addressing it in this PR, since I had to do a rebase anyways.

File: torch/csrc/distributed/rpc/process_group_agent.h
Comment (rohan-varma) [reply torch/csrc/distributed/rpc/process_group_agent.h]: Ended up doing it in this PR since had to do a rebase anyways.

**Pull request closed without merging**
